<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNAmodR.ML: detecting patterns of post-transcriptional modifications using machine learning • RNAmodR.ML</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RNAmodR.ML: detecting patterns of post-transcriptional modifications using machine learning">
<meta property="og:description" content="RNAmodR.ML">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RNAmodR.ML</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RNAmodR.ML.html">RNAmodR.ML: detecting patterns of post-transcriptional modifications using machine learning</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/RNAmodR.ML/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="RNAmodR.ML_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNAmodR.ML: detecting patterns of post-transcriptional modifications using machine learning</h1>
                        <h4 class="author">Felix G.M. Ernst and Denis L.J. Lafontaine</h4>
            
            <h4 class="date">2020-07-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/RNAmodR.ML/blob/master/vignettes/RNAmodR.ML.Rmd"><code>vignettes/RNAmodR.ML.Rmd</code></a></small>
      <div class="hidden name"><code>RNAmodR.ML.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Post-transcriptional modifications can be found abundantly in rRNA and tRNA and can be detected classically via several strategies. However, difficulties arise if the identity and the position of the modified nucleotides is to be determined at the same time. Classically, a primer extension, a form of reverse transcription (RT), would allow certain modifications to be accessed by blocks during the RT changes or changes in the cDNA sequences. Other modification would need to be selectively treated by chemical reactions to influence the outcome of the reverse transcription.</p>
<p>With the increased availability of high throughput sequencing, these classical methods were adapted to high throughput methods allowing more RNA molecules to be accessed at the same time. However, patterns of some modification cannot be detected by accessing small number of parameters. For these cases machine learning models can be trained on data from positions known to be modified in order to detect additional modified positions.</p>
<p>To extend the functionality of the <code>RNAmodR</code> package and classical detection strategies used for RiboMethSeq or AlkAnilineSeq (see <code>RNAmodR.RiboMethSeq</code> and <code>RNAmodR.AlkAnilineSeq</code> packages) towards detection through machine learning models, <code>RNAmodR.ML</code> provides classes and an example workflow.</p>
<p><a name="RNAmodR.ML"></a></p>
</div>
<div id="using-rnamodr-ml" class="section level1">
<h1 class="hasAnchor">
<a href="#using-rnamodr-ml" class="anchor"></a>Using RNAmodR.ML</h1>
<pre><code>## Failed to copy the script/BiocCheck or script/BiocCheckGitClone script</code></pre>
<pre><code>## to /opt/R/4.0.2/lib/R/bin. If you want to be able to run 'R CMD</code></pre>
<pre><code>## BiocCheck' you'll need to copy it yourself to a directory on your PATH,</code></pre>
<pre><code>## making sure it is executable. See the BiocCheck vignette for more</code></pre>
<pre><code>## information.</code></pre>
<pre><code>## using temporary cache /tmp/RtmpUeAGzy/BiocFileCache</code></pre>
<pre><code>## snapshotDate(): 2020-07-10</code></pre>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">GenomicRanges</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">RNAmodR.ML</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">RNAmodR.Data</span>)</pre></body></html></div>
<p>The <code>ModifierML</code> class extends the <code>Modifier</code> class from the <code>RNAmodR</code> package and adds one slot, <code>mlModel</code>, a getter/setter <code>getMLModel</code>/<code>setMLModel</code>, an additional <code>useMLModel</code> function to be called from the <code>aggregate</code> function.</p>
<p>The slot <code>mlModel</code> can either be an empty character or contain a name of a <code>ModifierMLModel</code> class, which is loaded upon creation of a <code>ModifierML</code> object, and serves as a wrapper around a machine learning model. For different types of machine learning models <code>ModifierMLModel</code> derived classes are available, which currently are:</p>
<ul>
<li>
<code>ModifierMLranger</code> for models generated with the <code>ranger</code> package <a href="#References"><span class="citation">(Wright and Ziegler 2017)</span></a>
</li>
<li>
<code>ModifierMLkeras</code> for models generated with the <code>keras</code> package <a href="#References"><span class="citation">(Allaire and Chollet 2018)</span></a>
</li>
</ul>
<p>To illustrate how to develop a machine learning model with help from the <code>RNAmodR.ML</code> package, an example is given below.</p>
</div>
<div id="development-of-new-modifier-class" class="section level1">
<h1 class="hasAnchor">
<a href="#development-of-new-modifier-class" class="anchor"></a>Development of new <code>Modifier</code> class</h1>
<p>As an example for this vignette, we will try to detect D positions in AlkAnilineSeq data. First define a specific <code>ModifierML</code> class loading pileup and coverage data. In this example, the RNA specific <code>RNAModifierML</code> class is used.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu">setClass</span>(<span class="st">"ModMLExample"</span>,
         <span class="kw">contains</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"RNAModifierML"</span>),
         <span class="kw">prototype</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">mod</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"D"</span>),
                          <span class="kw">score</span> <span class="kw">=</span> <span class="st">"score"</span>,
                          <span class="kw">dataType</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PileupSequenceData"</span>,
                                       <span class="st">"CoverageSequenceData"</span>),
                          <span class="kw">mlModel</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="fl">0</span>)))
<span class="co"># constructor function for ModMLExample</span>
<span class="no">ModMLExample</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">x</span>, <span class="no">annotation</span> <span class="kw">=</span> <span class="fl">NA</span>, <span class="no">sequences</span> <span class="kw">=</span> <span class="fl">NA</span>, <span class="no">seqinfo</span> <span class="kw">=</span> <span class="fl">NA</span>,
                           <span class="no">...</span>){
  <span class="kw pkg">RNAmodR</span><span class="kw ns">:::</span><span class="fu"><a href="https://rdrr.io/pkg/RNAmodR/man/Modifier-class.html">Modifier</a></span>(<span class="st">"ModMLExample"</span>, <span class="kw">x</span> <span class="kw">=</span> <span class="no">x</span>, <span class="kw">annotation</span> <span class="kw">=</span> <span class="no">annotation</span>,
                     <span class="kw">sequences</span> <span class="kw">=</span> <span class="no">sequences</span>, <span class="kw">seqinfo</span> <span class="kw">=</span> <span class="no">seqinfo</span>, <span class="no">...</span>)
}

<span class="fu">setClass</span>(<span class="st">"ModSetMLExample"</span>,
         <span class="kw">contains</span> <span class="kw">=</span> <span class="st">"ModifierSet"</span>,
         <span class="kw">prototype</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">elementType</span> <span class="kw">=</span> <span class="st">"ModMLExample"</span>))
<span class="co"># constructor function for ModSetMLExample</span>
<span class="no">ModSetMLExample</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">x</span>, <span class="no">annotation</span> <span class="kw">=</span> <span class="fl">NA</span>, <span class="no">sequences</span> <span class="kw">=</span> <span class="fl">NA</span>, <span class="no">seqinfo</span> <span class="kw">=</span> <span class="fl">NA</span>,
                              <span class="no">...</span>){
  <span class="kw pkg">RNAmodR</span><span class="kw ns">:::</span><span class="fu"><a href="https://rdrr.io/pkg/RNAmodR/man/ModifierSet-class.html">ModifierSet</a></span>(<span class="st">"ModMLExample"</span>, <span class="no">x</span>, <span class="kw">annotation</span> <span class="kw">=</span> <span class="no">annotation</span>,
                        <span class="kw">sequences</span> <span class="kw">=</span> <span class="no">sequences</span>, <span class="kw">seqinfo</span> <span class="kw">=</span> <span class="no">seqinfo</span>, <span class="no">...</span>)
}</pre></body></html></div>
<p>Since the <code>mlModel</code> slot contains an empty character, the creation of the object will not automatically trigger a search for modifications. However, it will aggregate the data in the format we want to use. The <code>aggregate_example</code> function is just an example and the aggregation of the data is part of the model building. (See (#Summary))</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu">setMethod</span>(
  <span class="kw">f</span> <span class="kw">=</span> <span class="st">"aggregateData"</span>,
  <span class="kw">signature</span> <span class="kw">=</span> <span class="fu">signature</span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">"ModMLExample"</span>),
  <span class="kw">definition</span> <span class="kw">=</span>
    <span class="kw">function</span>(<span class="no">x</span>){
      <span class="fu"><a href="../reference/RNAmodR.ML-example.html">aggregate_example</a></span>(<span class="no">x</span>)
    }
)</pre></body></html></div>
</div>
<div id="getting-training-data" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-training-data" class="anchor"></a>Getting training data</h1>
<p>To gather training data, we just create a <code>ModMLExample</code> object and let it do its job.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">me</span> <span class="kw">&lt;-</span>  <span class="fu">ModMLExample</span>(<span class="no">files</span><span class="kw">[[</span><span class="fl">1</span>]], <span class="no">annotation</span>, <span class="no">sequences</span>)</pre></body></html></div>
<pre><code>## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ... OK
## Loading Pileup data from BAM files ... OK
## Loading Coverage data from BAM files ... OK
## Aggregating data and calculating scores ... Starting to search for 'Dihydrouridine' ...</code></pre>
<pre><code>## Warning: ML model not set. Skipped search for modifications ...</code></pre>
<pre><code>## done.</code></pre>
<p>Afterwards we need to load/create coordinates for positions known to be modified as well as positions known to be unmodified.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dmod"</span>,<span class="kw">package</span> <span class="kw">=</span> <span class="st">"RNAmodR.ML"</span>)
<span class="co"># we just select the next U position from known positions</span>
<span class="no">nextUPos</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">gr</span>){
  <span class="no">nextU</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq.int</a></span>(<span class="fl">1L</span>,<span class="fl">2L</span>),
                  <span class="kw">function</span>(<span class="no">i</span>){
                    <span class="no">subseq</span> <span class="kw">&lt;-</span> <span class="fu">subseq</span>(<span class="fu">sequences</span>(<span class="no">me</span>)[<span class="no">dmod</span>$<span class="no">Parent</span>], <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">dmod</span>)+<span class="fl">3L</span>)
                    <span class="no">pos</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">dmod</span>) + <span class="fl">2L</span> +
                      <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">subseq</span>),<span class="st">""</span>),
                    <span class="kw">function</span>(<span class="no">y</span>){<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">y</span> <span class="kw">==</span> <span class="st">"U"</span>)[<span class="no">i</span>]},<span class="fu"><a href="https://rdrr.io/r/base/integer.html">integer</a></span>(<span class="fl">1</span>))
                    <span class="no">ans</span> <span class="kw">&lt;-</span> <span class="no">dmod</span>
                    <span class="fu">ranges</span>(<span class="no">ans</span>) <span class="kw">&lt;-</span> <span class="fu">IRanges</span>(<span class="kw">start</span> <span class="kw">=</span> <span class="no">pos</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">1L</span>)
                    <span class="no">ans</span>
                  })
  <span class="no">nextU</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">c</span>,<span class="no">nextU</span>)
  <span class="no">nextU</span>$<span class="no">mod</span> <span class="kw">&lt;-</span> <span class="kw">NULL</span>
  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">nextU</span>)
}
<span class="no">nondmod</span> <span class="kw">&lt;-</span> <span class="fu">nextUPos</span>(<span class="no">dmod</span>)
<span class="no">nondmod</span> <span class="kw">&lt;-</span> <span class="no">nondmod</span>[!(<span class="no">nondmod</span> <span class="kw">%in%</span> <span class="no">dmod</span>)]
<span class="no">coord</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">dmod</span>,<span class="no">nondmod</span>))
<span class="no">coord</span> <span class="kw">&lt;-</span> <span class="no">coord</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="no">coord</span>$<span class="no">Parent</span>))]</pre></body></html></div>
<p>With these coordinates the aggregated data of the <code>ModMLExample</code> can be subset to a training data set using the function <code>trainingData</code>.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">trainingData</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/trainingData.html">trainingData</a></span>(<span class="no">me</span>,<span class="no">coord</span>)
<span class="no">trainingData</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">trainingData</span>, <span class="kw">use.names</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="co"># converting logical labels to integer</span>
<span class="no">trainingData</span>$<span class="no">labels</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="no">trainingData</span>$<span class="no">labels</span>)</pre></body></html></div>
</div>
<div id="training-a-model" class="section level1">
<h1 class="hasAnchor">
<a href="#training-a-model" class="anchor"></a>Training a model</h1>
<p>How a specific model can be trained or what kind of strategies can be employed to successfully train a model, is out of scope for the vignette. For this example a random forest is trained using the functionality provided by the <code>ranger</code> package.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ranger</span>)
<span class="no">model</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ranger/man/ranger.html">ranger</a></span>(<span class="no">labels</span> ~ <span class="no">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">trainingData</span>))</pre></body></html></div>
</div>
<div id="constructing-a-modifiermlmodel" class="section level1">
<h1 class="hasAnchor">
<a href="#constructing-a-modifiermlmodel" class="anchor"></a>Constructing a ‘ModifierMLModel’</h1>
<p>Now, the trained model can be used to create a new <code>ModifierMLModel</code> class and object.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu">setClass</span>(<span class="st">"ModifierMLexample"</span>,
         <span class="kw">contains</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ModifierMLranger"</span>),
         <span class="kw">prototype</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">model</span> <span class="kw">=</span> <span class="no">model</span>))
<span class="no">ModifierMLexample</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">...</span>){
  <span class="fu">new</span>(<span class="st">"ModifierMLexample"</span>)
}
<span class="no">mlmodel</span> <span class="kw">&lt;-</span> <span class="fu">ModifierMLexample</span>()</pre></body></html></div>
<p>To be able to use the model via the <code>ModifierMLModel</code> class, we also need to define an accessor to the predictions made by the model. This function is called <code>useModel</code> and is already prefined for the <code>ModifierMLModel</code> classes mentioned above in secion <a href="#RNAmodR.ML">Using RNAmodR.ML</a>.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu">getMethod</span>(<span class="st">"useModel"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ModifierMLranger"</span>,<span class="st">"ModifierML"</span>))</pre></body></html></div>
<pre><code>## Method Definition:
## 
## function (x, y) 
## {
##     data &lt;- getAggregateData(y)
##     model &lt;- x@model
##     if (!is(model, "ranger")) {
##         stop("model is not a ranger model")
##     }
##     unlisted_data &lt;- unlist(data, use.names = FALSE)
##     p &lt;- predict(x@model, data.frame(unlisted_data))
##     relist(p$predictions, data)
## }
## &lt;bytecode: 0x5647dbe16f70&gt;
## &lt;environment: namespace:RNAmodR.ML&gt;
## 
## Signatures:
##         x                  y           
## target  "ModifierMLranger" "ModifierML"
## defined "ModifierMLranger" "ModifierML"</code></pre>
<p>If the results of these function is not usable for a specific model, it can redefined for your needs. As defined by <code>RNAmodR.ML</code> the function returns a <code>NumericList</code> along the aggregated data of the <code>ModifierML</code> object.</p>
</div>
<div id="setting-and-using-the-model" class="section level1">
<h1 class="hasAnchor">
<a href="#setting-and-using-the-model" class="anchor"></a>Setting and using the model</h1>
<p>The generated <code>ModifierMLexample</code> wrapper can now be set for the <code>ModifierML</code> object using the <code>setMLModel</code> function. (If a model is saved as part of a package, this step ist not necessary, because it can be part of the class definition)</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu">setMLModel</span>(<span class="no">me</span>) <span class="kw">&lt;-</span> <span class="no">mlmodel</span></pre></body></html></div>
<p>In order for the prediction data to be usable, another function has to be implemented to save the predictions in the aggregated data. The function is called <code>useMLModel</code>.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu">setMethod</span>(<span class="kw">f</span> <span class="kw">=</span> <span class="st">"useMLModel"</span>,
          <span class="kw">signature</span> <span class="kw">=</span> <span class="fu">signature</span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">"ModMLExample"</span>),
          <span class="kw">definition</span> <span class="kw">=</span>
            <span class="kw">function</span>(<span class="no">x</span>){
              <span class="no">predictions</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/ModifierMLModel-class.html">useModel</a></span>(<span class="fu"><a href="../reference/ModifierML-class.html">getMLModel</a></span>(<span class="no">x</span>), <span class="no">x</span>)
              <span class="no">data</span> <span class="kw">&lt;-</span> <span class="fu">getAggregateData</span>(<span class="no">x</span>)
              <span class="no">unlisted_data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">data</span>, <span class="kw">use.names</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
              <span class="no">unlisted_data</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">predictions</span>)
              <span class="no">x</span>@<span class="kw">aggregate</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/relist.html">relist</a></span>(<span class="no">unlisted_data</span>,<span class="no">data</span>)
              <span class="no">x</span>
            }
)</pre></body></html></div>
<p>By re-running the <code>aggregate</code> function and force an update of the data, the predictions are made and used to populate the <code>score</code> column as outlined above.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">me</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span>(<span class="no">me</span>, <span class="kw">force</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
</div>
<div id="performance" class="section level1">
<h1 class="hasAnchor">
<a href="#performance" class="anchor"></a>Performance</h1>
<p>During the model building phase some form of a performance measurement usually is included. In addition to these model specific measurements, <code>RNAmodR.ML</code> includes the functionality from the <code>ROCR</code> package inherited from the <code>RNAmodR</code> package. With this the performance of a model can evaluted over the training set or any coordinates.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu">plotROC</span>(<span class="no">me</span>, <span class="no">dmod</span>)</pre></body></html></div>
<div class="figure">
<img src="RNAmodR.ML_files/figure-html/plot1-1.png" alt="Performance of the maching learning model to distinguish unmodified from modified nucleotides." width="700"><p class="caption">
Performance of the maching learning model to distinguish unmodified from modified nucleotides.
</p>
</div>
</div>
<div id="using-a-modifierml-class" class="section level1">
<h1 class="hasAnchor">
<a href="#using-a-modifierml-class" class="anchor"></a>Using a <code>ModifierML</code> class</h1>
<p>Since we want to use the <code>ModifierML</code> object to detect modifications, we also need to define the <code>findMod</code> function. Based on the information on the performance, we set a threshold of <code>0.8</code> for the prediction score for detecting D modifications. In the example below this threshold is hardcoded in the <code>find_mod_example</code> function, but can also be implemented using the <code>settings</code> function.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="fu">setMethod</span>(
  <span class="kw">f</span> <span class="kw">=</span> <span class="st">"findMod"</span>,
  <span class="kw">signature</span> <span class="kw">=</span> <span class="fu">signature</span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">"ModMLExample"</span>),
  <span class="kw">definition</span> <span class="kw">=</span>
    <span class="kw">function</span>(<span class="no">x</span>){
      <span class="fu"><a href="../reference/RNAmodR.ML-example.html">find_mod_example</a></span>(<span class="no">x</span>, <span class="fl">25L</span>)
    }
)</pre></body></html></div>
<p>Now we can redfine the <code>ModMLExample</code> class with the slot <code>mlModel</code> already set to the <code>ModifierMLexample</code> class. The <code>ModMLExample</code> is now complete.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span>(<span class="no">me</span>)
<span class="fu">setClass</span>(<span class="st">"ModMLExample"</span>,
         <span class="kw">contains</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"RNAModifierML"</span>),
         <span class="kw">prototype</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">mod</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"D"</span>),
                          <span class="kw">score</span> <span class="kw">=</span> <span class="st">"score"</span>,
                          <span class="kw">dataType</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PileupSequenceData"</span>,
                                       <span class="st">"CoverageSequenceData"</span>),
                          <span class="kw">mlModel</span> <span class="kw">=</span> <span class="st">"ModifierMLexample"</span>))
<span class="no">me</span> <span class="kw">&lt;-</span>  <span class="fu">ModMLExample</span>(<span class="no">files</span><span class="kw">[[</span><span class="fl">1</span>]], <span class="no">annotation</span>, <span class="no">sequences</span>)</pre></body></html></div>
<pre><code>## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ... OK
## Loading Pileup data from BAM files ... OK
## Loading Coverage data from BAM files ... OK
## Aggregating data and calculating scores ... Starting to search for 'Dihydrouridine' ... done.</code></pre>
<p>The detected modifications can be access using the <code>modifications</code> function.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">mod</span> <span class="kw">&lt;-</span> <span class="fu">modifications</span>(<span class="no">me</span>)
<span class="no">mod</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="no">mod</span>, <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">mod</span>$<span class="no">Parent</span>,<span class="kw">levels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">mod</span>$<span class="no">Parent</span>)))
<span class="no">mod</span></pre></body></html></div>
<pre><code>## GRangesList object of length 35:
## $`1`
## GRanges object with 2 ranges and 5 metadata columns:
##             seqnames    ranges strand |         mod      source        type
##                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;
##   [1] Q0020_15S_RRNA        48      + |           D  RNAmodR.ML      RNAMOD
##   [2] Q0020_15S_RRNA       323      + |           D  RNAmodR.ML      RNAMOD
##           score      Parent
##       &lt;numeric&gt; &lt;character&gt;
##   [1]  0.932482           1
##   [2]  0.836582           1
##   -------
##   seqinfo: 60 sequences from an unspecified genome; no seqlengths
## 
## $`3`
## GRanges object with 2 ranges and 5 metadata columns:
##       seqnames    ranges strand |         mod      source        type     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
##   [1]  RDN18-1       229      + |           D  RNAmodR.ML      RNAMOD    0.8251
##   [2]  RDN18-1      1669      + |           D  RNAmodR.ML      RNAMOD    0.8188
##            Parent
##       &lt;character&gt;
##   [1]           3
##   [2]           3
##   -------
##   seqinfo: 60 sequences from an unspecified genome; no seqlengths
## 
## $`4`
## GRanges object with 4 ranges and 5 metadata columns:
##       seqnames    ranges strand |         mod      source        type     score
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;
##   [1]  RDN25-1         2      + |           D  RNAmodR.ML      RNAMOD  0.867700
##   [2]  RDN25-1       748      + |           D  RNAmodR.ML      RNAMOD  0.828033
##   [3]  RDN25-1      1720      + |           D  RNAmodR.ML      RNAMOD  0.835500
##   [4]  RDN25-1      2571      + |           D  RNAmodR.ML      RNAMOD  0.801367
##            Parent
##       &lt;character&gt;
##   [1]           4
##   [2]           4
##   [3]           4
##   [4]           4
##   -------
##   seqinfo: 60 sequences from an unspecified genome; no seqlengths
## 
## ...
## &lt;32 more elements&gt;</code></pre>
</div>
<div id="refining-a-model" class="section level1">
<h1 class="hasAnchor">
<a href="#refining-a-model" class="anchor"></a>Refining a model</h1>
<p>Some of the modification found look reasonable. However, some of the positions seem to be noise.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="kw">ucscChromosomeNames</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu">plotDataByCoord</span>(<span class="fu">sequenceData</span>(<span class="no">me</span>),<span class="no">mod</span><span class="kw">[[</span><span class="st">"4"</span>]][<span class="fl">1</span>])</pre></body></html></div>
<div class="figure">
<img src="RNAmodR.ML_files/figure-html/plot2-1.png" alt="Visualization of sequence data" width="700"><p class="caption">
Visualization of sequence data
</p>
</div>
<p>Several options exist to improve the model: Either the threshold applied to the prediction score can be raised to a higher value, like <code>0.9</code> or the model can maybe retrained by including these position in another training data set. In addition, the training data might be improved in general by higher sequencing depth.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">nonValidMod</span> <span class="kw">&lt;-</span> <span class="no">mod</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"1"</span>,<span class="st">"4"</span>)]
<span class="no">nonValidMod</span><span class="kw">[[</span><span class="st">"18"</span>]] <span class="kw">&lt;-</span> <span class="no">nonValidMod</span><span class="kw">[[</span><span class="st">"18"</span>]][<span class="fl">2</span>]
<span class="no">nonValidMod</span><span class="kw">[[</span><span class="st">"26"</span>]] <span class="kw">&lt;-</span> <span class="no">nonValidMod</span><span class="kw">[[</span><span class="st">"26"</span>]][<span class="fl">2</span>]
<span class="no">nonValidMod</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">nonValidMod</span>)
<span class="no">nonValidMod</span> <span class="kw">&lt;-</span> <span class="no">nonValidMod</span>[,<span class="st">"Parent"</span>]
<span class="no">coord</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">dmod</span>,<span class="no">nondmod</span>,<span class="no">nonValidMod</span>))
<span class="no">coord</span> <span class="kw">&lt;-</span> <span class="no">coord</span>[<span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="no">coord</span>$<span class="no">Parent</span>))]</pre></body></html></div>
<p>As an example, a new model is trained including the wrongly identified positions from the previous model as unmodified positions.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">trainingData</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/trainingData.html">trainingData</a></span>(<span class="no">me</span>,<span class="no">coord</span>)
<span class="no">trainingData</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">trainingData</span>, <span class="kw">use.names</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">trainingData</span>$<span class="no">labels</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="no">trainingData</span>$<span class="no">labels</span>)</pre></body></html></div>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">model2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ranger/man/ranger.html">ranger</a></span>(<span class="no">labels</span> ~ <span class="no">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">trainingData</span>), <span class="kw">num.trees</span> <span class="kw">=</span> <span class="fl">2000</span>)
<span class="fu">setClass</span>(<span class="st">"ModifierMLexample2"</span>,
         <span class="kw">contains</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ModifierMLranger"</span>),
         <span class="kw">prototype</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">model</span> <span class="kw">=</span> <span class="no">model2</span>))
<span class="no">ModifierMLexample2</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">...</span>){
  <span class="fu">new</span>(<span class="st">"ModifierMLexample2"</span>)
}
<span class="no">mlmodel2</span> <span class="kw">&lt;-</span> <span class="fu">ModifierMLexample2</span>()
<span class="no">me2</span> <span class="kw">&lt;-</span> <span class="no">me</span>
<span class="fu">setMLModel</span>(<span class="no">me2</span>) <span class="kw">&lt;-</span> <span class="no">mlmodel2</span>
<span class="no">me2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span>(<span class="no">me2</span>, <span class="kw">force</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>After updating the <code>ModifierMLexample</code> class and aggregating the data again the performance looks a bit better…</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="fu">plotROC</span>(<span class="no">me2</span>, <span class="no">dmod</span>, <span class="kw">score</span><span class="kw">=</span><span class="st">"score"</span>)</pre></body></html></div>
<div class="figure">
<img src="RNAmodR.ML_files/figure-html/plot3-1.png" alt="Performance aggregation of multiple samples and strategies." width="700"><p class="caption">
Performance aggregation of multiple samples and strategies.
</p>
</div>
<p>… and leads to a better result for detecting D modifications. Some positions are not detected anymore, which suggest that this model is still not an optimal solution and other factors could and should be improved upon as suggested above.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu">setMethod</span>(
  <span class="kw">f</span> <span class="kw">=</span> <span class="st">"findMod"</span>,
  <span class="kw">signature</span> <span class="kw">=</span> <span class="fu">signature</span>(<span class="kw">x</span> <span class="kw">=</span> <span class="st">"ModMLExample"</span>),
  <span class="kw">definition</span> <span class="kw">=</span>
    <span class="kw">function</span>(<span class="no">x</span>){
      <span class="fu"><a href="../reference/RNAmodR.ML-example.html">find_mod_example</a></span>(<span class="no">x</span>, <span class="fl">25L</span>)
    }
)
<span class="no">me2</span> <span class="kw">&lt;-</span> <span class="fu">modify</span>(<span class="no">me2</span>, <span class="kw">force</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu">modifications</span>(<span class="no">me2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 44 ranges and 5 metadata columns:
##         seqnames    ranges strand |         mod      source        type
##            &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;character&gt;
##    [1]  tA-AGC-D        47      + |           D  RNAmodR.ML      RNAMOD
##    [2]  tA-TGC-A        47      + |           D  RNAmodR.ML      RNAMOD
##    [3]  tC-GCA-B        19      + |           D  RNAmodR.ML      RNAMOD
##    [4]  tC-GCA-B        46      + |           D  RNAmodR.ML      RNAMOD
##    [5]  tE-CTC-D        20      + |           D  RNAmodR.ML      RNAMOD
##    ...       ...       ...    ... .         ...         ...         ...
##   [40]  tV-AAC-O        48      + |           D  RNAmodR.ML      RNAMOD
##   [41]  tV-CAC-D        47      + |           D  RNAmodR.ML      RNAMOD
##   [42] tW-CCA-G1        16      + |           D  RNAmodR.ML      RNAMOD
##   [43] tW-CCA-G1        46      + |           D  RNAmodR.ML      RNAMOD
##   [44]  tY-GTA-D        21      + |           D  RNAmodR.ML      RNAMOD
##            score      Parent
##        &lt;numeric&gt; &lt;character&gt;
##    [1]  0.966675           6
##    [2]  0.950492           7
##    [3]  0.915256           8
##    [4]  0.937725           8
##    [5]  0.827975          11
##    ...       ...         ...
##   [40]  0.929950          56
##   [41]  0.926875          57
##   [42]  0.811967          59
##   [43]  0.831342          59
##   [44]  0.913658          60
##   -------
##   seqinfo: 60 sequences from an unspecified genome; no seqlengths</code></pre>
<p>In addition to training a single model, several models can be trained and combined to a <code>ModifierSet</code>.</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">mse</span> <span class="kw">&lt;-</span> <span class="fu">ModSetMLExample</span>(<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">one</span> <span class="kw">=</span> <span class="no">me</span>, <span class="kw">two</span> <span class="kw">=</span> <span class="no">me2</span>))</pre></body></html></div>
<p>An overall performance over several models can be analyzed or the individual performance compaired.</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="fu">plotROC</span>(<span class="no">mse</span>, <span class="no">dmod</span>, <span class="kw">score</span><span class="kw">=</span> <span class="st">"score"</span>,
        <span class="kw">plot.args</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">avg</span> <span class="kw">=</span> <span class="st">"threshold"</span>, <span class="kw">spread.estimate</span> <span class="kw">=</span> <span class="st">"stderror"</span>))</pre></body></html></div>
<div class="figure">
<img src="RNAmodR.ML_files/figure-html/plot4-1.png" alt="Performance average across models" width="700"><p class="caption">
Performance average across models
</p>
</div>
<p>If several models are trained and each provides useful information, these can be package into a single <code>ModifierMLModel</code> class to combine the output of several models. Some of the functions outlined above need, e.g. <code>useMLModel</code> and/or <code>useModel</code>, to be modified to provide one or more scores for detecting a modification.</p>
</div>
<div id="packaging" class="section level1">
<h1 class="hasAnchor">
<a href="#packaging" class="anchor"></a>Packaging</h1>
<p>If the created models can be saved to file, they can also be included in a package. This would allow others to use the models and the models can potentially be improved upon with increasing version numbers.</p>
<p><a name="Summary"></a></p>
</div>
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<p><code>RNAmodR.ML</code> provides the interface for machine learning models to be used with <code>RNAmodR</code> to detect modified nucleotides in high throughput sequencing data. For your own work defining a working model might take some time. We hope that by using <code>RNAmodR.ML</code> the steps surrounding this crucial step might become a bit easier.</p>
<p>However, if some steps or design choices made for <code>RNAmodR.ML</code> do not suit your need, let us know. Contributions are always welcome as well.</p>
</div>
<div id="hints" class="section level1">
<h1 class="hasAnchor">
<a href="#hints" class="anchor"></a>Hints</h1>
<p>We also want to provide some additional hints and suggestions for developing machine learning models.</p>
<ol style="list-style-type: decimal">
<li>the aggregate function is used in the example above as a feature engineering tool. You might want to skip this step, if you want to use a deep learning model for example with <code>keras</code>.</li>
<li>If you don’t want to engage in a feature enginering step and just want to aggregate the sequence data as is, just do a custom <code>cbind</code> on the data from the <code>SequenceData</code> objects (<code>cbind</code> works on <code>SequenceData</code> objects, if they are of the same type. Convert each of them to a <code>CompressedSplitDataFrameList</code> using <code>as(x,"CompressedSplitDataFrameList")</code>).</li>
<li>in a deep learning context, if a coordinate is selected without a flanking value, e.g. when using <code>trainingData</code>, a 2D tensor is returned (sample, values). This can be converted into a 3D tensor by providing a flanking value.</li>
</ol>
</div>
<div id="sessioninfo" class="section level1">
<h1 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>Sessioninfo</h1>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ranger_0.12.1            Rsamtools_2.5.3          RNAmodR.Data_1.3.0      
##  [4] ExperimentHubData_1.15.0 AnnotationHubData_1.19.2 futile.logger_1.4.3     
##  [7] ExperimentHub_1.15.0     AnnotationHub_2.21.1     BiocFileCache_1.13.0    
## [10] dbplyr_1.4.4             RNAmodR.ML_1.3.1         RNAmodR_1.3.4           
## [13] Modstrings_1.5.1         Biostrings_2.57.2        XVector_0.29.3          
## [16] rtracklayer_1.49.3       GenomicRanges_1.41.5     GenomeInfoDb_1.25.8     
## [19] IRanges_2.23.10          S4Vectors_0.27.12        BiocGenerics_0.35.4     
## [22] BiocStyle_2.17.0        
## 
## loaded via a namespace (and not attached):
##   [1] backports_1.1.8               Hmisc_4.4-0                  
##   [3] plyr_1.8.6                    lazyeval_0.2.2               
##   [5] splines_4.0.2                 BiocParallel_1.23.2          
##   [7] ggplot2_3.3.2                 digest_0.6.25                
##   [9] ensembldb_2.13.1              htmltools_0.5.0              
##  [11] gdata_2.18.0                  magrittr_1.5                 
##  [13] checkmate_2.0.0               memoise_1.1.0                
##  [15] BSgenome_1.57.4               cluster_2.1.0                
##  [17] ROCR_1.0-11                   matrixStats_0.56.0           
##  [19] askpass_1.1                   pkgdown_1.5.1                
##  [21] prettyunits_1.1.1             jpeg_0.1-8.1                 
##  [23] colorspace_1.4-1              blob_1.2.1                   
##  [25] rappdirs_0.3.1                xfun_0.15                    
##  [27] dplyr_1.0.0                   jsonlite_1.7.0               
##  [29] crayon_1.3.4                  RCurl_1.98-1.2               
##  [31] graph_1.67.1                  survival_3.1-12              
##  [33] VariantAnnotation_1.35.3      glue_1.4.1                   
##  [35] gtable_0.3.0                  zlibbioc_1.35.0              
##  [37] DelayedArray_0.15.7           scales_1.1.1                 
##  [39] futile.options_1.0.1          DBI_1.1.0                    
##  [41] Rcpp_1.0.5                    xtable_1.8-4                 
##  [43] progress_1.2.2                htmlTable_2.0.1              
##  [45] foreign_0.8-80                bit_1.1-15.2                 
##  [47] OrganismDbi_1.31.0            Formula_1.2-3                
##  [49] AnnotationForge_1.31.2        getopt_1.20.3                
##  [51] htmlwidgets_1.5.1             httr_1.4.2                   
##  [53] gplots_3.0.4                  RColorBrewer_1.1-2           
##  [55] acepack_1.4.1                 ellipsis_0.3.1               
##  [57] pkgconfig_2.0.3               XML_3.99-0.4                 
##  [59] Gviz_1.33.2                   nnet_7.3-14                  
##  [61] tidyselect_1.1.0              rlang_0.4.7                  
##  [63] reshape2_1.4.4                later_1.1.0.1                
##  [65] AnnotationDbi_1.51.1          munsell_0.5.0                
##  [67] biocViews_1.57.3              BiocVersion_3.12.0           
##  [69] tools_4.0.2                   generics_0.0.2               
##  [71] RSQLite_2.2.0                 evaluate_0.14                
##  [73] stringr_1.4.0                 fastmap_1.0.1                
##  [75] yaml_2.2.1                    knitr_1.29                   
##  [77] bit64_0.9-7.1                 fs_1.4.2                     
##  [79] caTools_1.18.0                purrr_0.3.4                  
##  [81] AnnotationFilter_1.13.0       RBGL_1.65.0                  
##  [83] mime_0.9                      formatR_1.7                  
##  [85] biomaRt_2.45.2                compiler_4.0.2               
##  [87] rstudioapi_0.11               curl_4.3                     
##  [89] png_0.1-7                     interactiveDisplayBase_1.27.5
##  [91] tibble_3.0.3                  stringi_1.4.6                
##  [93] highr_0.8                     GenomicFeatures_1.41.2       
##  [95] desc_1.2.0                    lattice_0.20-41              
##  [97] ProtGenerics_1.21.0           Matrix_1.2-18                
##  [99] vctrs_0.3.2                   stringdist_0.9.6             
## [101] BiocCheck_1.25.4              pillar_1.4.6                 
## [103] lifecycle_0.2.0               optparse_1.6.6               
## [105] RUnit_0.4.32                  BiocManager_1.30.10          
## [107] data.table_1.12.8             bitops_1.0-6                 
## [109] httpuv_1.5.4                  colorRamps_2.3               
## [111] R6_2.4.1                      latticeExtra_0.6-29          
## [113] bookdown_0.20                 promises_1.1.1               
## [115] KernSmooth_2.23-17            gridExtra_2.3                
## [117] codetools_0.2-16              lambda.r_1.2.4               
## [119] dichromat_2.0-0               gtools_3.8.2                 
## [121] MASS_7.3-51.6                 assertthat_0.2.1             
## [123] SummarizedExperiment_1.19.6   openssl_1.4.2                
## [125] rprojroot_1.3-2               rBiopaxParser_2.29.1         
## [127] GenomicAlignments_1.25.3      GenomeInfoDbData_1.2.3       
## [129] hms_0.5.3                     grid_4.0.2                   
## [131] rpart_4.1-15                  rmarkdown_2.3                
## [133] biovizBase_1.37.0             Biobase_2.49.0               
## [135] shiny_1.5.0                   base64enc_0.1-3</code></pre>
<p><a name="References"></a></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Allaire-2018">
<p>Allaire, JJ, and François Chollet. 2018. <em>Keras: R Interface to ’Keras’</em>. <a href="https://CRAN.R-project.org/package=keras">https://CRAN.R-project.org/package=keras</a>.</p>
</div>
<div id="ref-Wright-2017">
<p>Wright, Marvin N., and Andreas Ziegler. 2017. “ranger: A Fast Implementation of Random Forests for High Dimensional Data in C++ and R.” <em>Journal of Statistical Software</em> 77 (1): 1–17. <a href="https://doi.org/10.18637/jss.v077.i01">https://doi.org/10.18637/jss.v077.i01</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
